### PR TITLE
[7.4-only] [LPS-124169,LPS-123931] SPA Engine consolidation + Calendar SPA issue fix

### PR DESCRIPTION
**Pull request by jbalsas**
This PR contains:
- Fix for [LPS-123931](https://issues.liferay.com/browse/LPS-123931) which was caused by a bug in our `runScriptsInElement` function (cc @bryceosterhaus.. take a look when you get back, not sure this covers everything, but at least it seems to fix Calendar.
- Many of the small improvements @wincent has requested over the past few weeks as while reviewing our integration of Senna.js inside the DXP codebase. I've left as much information as possible in the commit messages

Here's a list of the changes in the PR. I tried explaining each move in the commit message in detail:
- Simplify initialization logic
- Remove unnecessary utils class
- Inline errors module
- Remove globals module
- Remove version
- Remove data attribute initialization support
- Remove unnecessary `'use strict'` pragma
- Remove unnecessary configuration overrides
- Remove unnecessary .es prefixes
- Merge `LiferaySurface` into `Surface`
- Export all utils from the same module
- Consolidate `liferay` and `senna` root directories

**Test Plan**
- The module's `frontend-js-spa-web` unit tests are green
- Manual inspection of navigation looks good
- CI should be green 🤞

### CODE REVIEW

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by wincent** [comment]
> * Fix for [LPS-123931](https://issues.liferay.com/browse/LPS-123931) which was caused by a bug in our `runScriptsInElement` function

Seeing as I am lazy, which commit contains the fix? (not obvious by looking at the titles alone, and I haven't looked at the code yet)

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js [hunk 1]
```diff
@@ -376,9 +372,7 @@ class HtmlScreen extends RequestScreen {
 	 * @return {array.<Element>}
 	 */
 	querySelectorAll_(selector) {
-		return Array.prototype.slice.call(
-			globals.document.querySelectorAll(selector)
-		);
+		return Array.prototype.slice.call(document.querySelectorAll(selector));
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js:375]: Probably nicer as:

```js
return Array.from(document.querySelectorAll(selector));
```

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/RequestScreen.js [hunk 2]
```diff
@@ -14,11 +14,14 @@
...
-import errors from '../errors/errors';
-import globals from '../globals/globals';
-import {getUrlPath} from '../utils/utils';
+import {getUrlPath} from '../util/utils';
 import Screen from './Screen';
 
+const INVALID_STATUS = 'Invalid status code';
+const REQUEST_ERROR = 'Request error';
+const REQUEST_TIMEOUT = 'Request timeout';
+const REQUEST_PREMATURE_TERMINATION = 'Request terminated prematurely';
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/RequestScreen.js:23]: Probably should sort these.

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/surface/Surface.js [hunk 3]
```diff
@@ -90,27 +89,33 @@ class Surface extends Disposable {
 	 * @return {Element}
 	 */
 	addContent(screenId, opt_content) {
-		var child = this.defaultChild;
+		const fragment =
+			typeof opt_content === 'string'
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/surface/Surface.js:93]: If you can be bothered, now might be a nice time to rename `opt_content` to get rid of the underscore and switch to camel case (bonus points if you get rid of the "opt" abbreviation too).

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/surface/Surface.js [hunk 4]
```diff
@@ -125,7 +130,7 @@ class Surface extends Disposable {
 	 * @return {Element}
 	 */
 	createChild(screenId) {
-		var child = globals.document.createElement('div');
+		var child = document.createElement('div');
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/surface/Surface.js:133]: Mixed `var`/`const`/`let` is a bit distracting throughout all these files. Not necessary for this PR, but I'd put it on the "Boy Scout TODO List" for possible improvement.

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/util/utils.js [hunk 5]
```diff
@@ -215,6 +190,27 @@ export function removePathTrailingSlash(path) {
 	return path;
 }
 
+/**
+ * Destroys all rendered portlets on the page
+ */
+export function resetAllPortlets() {
+	getPortletBoundaryIds(Liferay.Portlet.list).forEach((value) => {
+		const portlet = document.querySelector('#' + value);
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/util/utils.js:198]: If you have an ID, [don't make a string-based selector](https://github.com/liferay/liferay-frontend-guidelines/issues/19). Just use `getElementById()`.

**Comment by jbalsas** [comment]
> LPS-123931 Makes sure script evaluation continues after finding irrel…

> > * Fix for [LPS-123931](https://issues.liferay.com/browse/LPS-123931) which was caused by a bug in our `runScriptsInElement` function
> 
> Seeing as I am lazy, which commit contains the fix? (not obvious by looking at the titles alone, and I haven't looked at the code yet)

[LPS-123931 Makes sure script evaluation continues after finding irrelevant script nodes](https://github.com/liferay-frontend/liferay-portal/pull/565/commits/5ca15fcc5a000498ed70f971012bc1bdfd3df4aa)

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
@wincent, there's a bunch of one-file-folders which should be trivial to get rid of, but didn't want to add more noise to an already big PR :)

File: modules/apps/frontend-js/frontend-js-spa-web/test/app/App.js [hunk 6]
```diff
@@ -125,64 +128,72 @@ describe('App', function () {
...
+			hash: '',
+			host: '',
+			hostname: '',
+			pathname: '/path',
+			search: '',
+		});
+
 		expect(this.app.canNavigate('/path#hashbang')).toBe(false);
+
+		restoreWindowLocation();
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/test/app/App.js:142]: If any test fails, `restoreWindowLocation()` won't run. Can we save in `beforeEach()` and restore in `afterEach()` instead?

**Review by wincent — COMMENTED** [review]
Lovely to see so much cleanup. If it's green (who am I kidding?), lg2m.

**Comment by wincent** [comment]
My only wish-list item now is to see the redundancy in the paths (eg. `util/utils`) go away.

**Comment by jbalsas** [comment]
Surprisingly enough, `ModifyPaginationPortalProperties ` might be legit... we don't navigate when clicking directly within the `svg` in the button. I'm not sure what I did here that could trigger it, so I'm going to retest just to make sure it's not flakey while I investigate further.

**Comment by jbalsas** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/LiferayApp.js [hunk 7]
```diff
@@ -146,17 +159,11 @@ class LiferayApp extends App {
 	 */
 
 	isInPortletBlacklist(element) {
-		return Object.keys(this.portletsBlacklist).some((portletId) => {
-			const boundaryId = getPortletBoundaryId(portletId);
+		const portletsBlacklistSelector = this.portletsBlacklist
+			.map((portletId) => `[id^="${getPortletBoundaryId(portletId)}"]`)
+			.join();
```
[the remark this answers is not in the document]
Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/LiferayApp.js:162]: it's not really that big... but that's a fair concern... 🤔

**Comment by jbalsas** [comment]
Okay, so I figured this was caused by https://github.com/liferay/liferay-portal/commit/5789a83372c639ab489fd3b3085e5e156c447161#diff-25174ea2f029c1856199a18ff842555bcf9bb92f1d47a1cad96a85115412e49c 

@wincent, this might also be causing the `data-restore-title` issues you mentioned in Slack before... 
@kresimir-coko, if this is green I plan on forwarding it and then we can investigate the change to our `delegate` in the context of `tooltip-support` which is in fact not doing so well ;)

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
@wincent, I'll address all of your concerns in another followup... trying to get this in today! 🤞

**Comment by kresimir-coko** [comment]
@jbalsas 
> Okay, so I figured this was caused by ...

There is a bug in that (my) PR, that we didn't catch until recently, caused by me replacing `this.eventListener_.removeAllListeners()` with `this.eventListener_.dispose()`. At the time I thought it would work, but that replacement only works when `this.eventListener_` is a variable that holds one `delegate` call.

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
ci:test:bundle

**Comment by jbalsas** [comment]
Hey @john-co, @PablitoBonito, @zhangyan0701 I can't reproduce this locally using the [bundle](https://files.liferay.com/private/bundles/test-1-14/jobs/test-portal-acceptance-pullrequest(master)/builds/7511/liferay-portal-bundle-tomcat.tar.gz).

The fact that it has failed twice makes me think it's a valid issue, but not being able to reproduce it manually, I'm leaning towards quarantining the test for now.

Could you please take a look and disable that test for now if nothing obvious makes sense?

**Comment by jbalsas** [comment]
Ok, never mind, I think I finally understand the issue... the test expects the page to refresh after changing the search configuration which is now not happening 🤔

**Comment by jbalsas** [comment]
I've tested it and it actually also doesn't work in `master`, so maybe that's not even it 😢

**Comment by john-co** [comment]
cc/ @xbrianlee , we're planning to quarantine the failing search test for now to push this PR through since we're not immediately able to tell what's the issue. Could you take a look when you get the chance?

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
@xbrianlee, see https://github.com/liferay-frontend/liferay-portal/pull/565/commits/b7c621601f3e081aba53b4bc8d2334f2ec50acb6 for the quarantined test.

I haven't managed to replicate the issue which if I understand correctly is doing:
- Create 3 WC
- Search to see the 3 of them
- Change Search Results delta to 2
- Close the Configuration window
- Somehow search again to refresh the page? (I guess this is implicit in the macro, because the results do not show after saving the configuration and require a full refresh of the page)
- Click on Next
- Assert that `WC Title 3` is in the next page

If that's what the test is doing, it seems to be working for me. 

At the risk of overstepping, I wonder why the extra navigation is needed since we already see the `delta` applied (in the sense that we only see 2 results now). Couldn't we simply count the amount of results to see that pagination configuration indeed takes effect and avoid making the test more complex?

**Comment by jbalsas** [comment]
ci:stop

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by wincent** [comment]
> There is a bug in that (my) PR, that we didn't catch until recently, caused by me replacing `this.eventHandler_.removeAllListeners()` with `this.eventHandler_.dispose()`. At the time I thought it would work, but that replacement only works when `this.eventHandler_` is a variable that holds one `delegate` call.

In the interests of precision for anybody driving by here an "event handler" in this context is an abstraction from `metal-events` that allows you to manage a bunch of event listeners as a group; you `add()` "disposables" to the handler and the handler has [a `removeAllListeners()` method](https://github.com/metal/metal.js/blob/5b5032dd840bc3364894776c8bb2fb82517e2553/packages/metal-events/src/EventHandler.js#L48-L54) that you can call to clean them all up at once. It does not have a `dispose()` method.

A variable does not "hold" a `delegate` call; `delegate()` _returns_ an object with a `dispose()` method that can be called to clean up the underlying event listener.

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by xbrianlee** [comment]
@jbalsas @john-co I was able to manually reproduce the issue in `Search#ModifyPaginationPortalProperties` in the bundle above with the following steps:
1. Set `search.container.page.default.delta=2` in portal.properties
2. Create 3 WC - WC Title 1, WC Title 2, WC Title3
3. Search for "WC Title" in the Search Bar
4. Paginate to next page of results

It's here that the page loads indefinitely for me. This seems related to the SPA changes in this PR. Can you guys take another look?

> At the risk of overstepping, I wonder why the extra navigation is needed since we already see the delta applied (in the sense that we only see 2 results now). Couldn't we simply count the amount of results to see that pagination configuration indeed takes effect and avoid making the test more complex?
@jbalsas Good point, actually we're testing that pagination works properly here with a non-default delta. We've had regressions in the past where everything in the Search Results UI will work as expected besides pagination and the easiest way to force an issue was to paginate.

**Comment by jbalsas** [comment]
Hey @xbrianlee, thanks for checking that out! 

After quite a while I have managed to reproduce using production settings only. It doesn't happen always but it is often enough to get an idea of what's going on.

I've already spent some time looking into this and there's a very weird interaction between:
- How we combine resources when `js_fast_load=1` 
- The order in which scripts execute in those conditions
- The time when `PersonalMenu` is created
- The time when `Senna` is initialized

Some type of race condition causes that, under some circumstances still to determine, we try to unmount a dropdown `ClayPortal` from a parent different to the one it was mounted on. That is, [we mount it on `document.body`](https://github.com/liferay/clay/blob/master/packages/clay-shared/src/Portal.tsx#L45) but then by the time we [try to unmount it](https://github.com/liferay/clay/blob/master/packages/clay-shared/src/Portal.tsx#L50), because we've changed the parent, the operation fails. 

Being a race condition, it's hard to determine what the correct fix should be so we still need more time to figure this out. It's also very likely this is already happening in other places in portal, so this is also a good chance to debug it since it happens more or less reliably.

**@john-co, could you please properly quarantine this test and make sure this gets in today? This way we unblock a good chunk of further work and can look at this now that the issue is more or less clear.**

Hope that's okay, @xbrianlee... we really need to get this in soon so we can continue stabilizing the platform. The more we drag this over the riskier it gets 🙏

**Comment by xbrianlee** [comment]
ci:test:search

(To catch other tests that may fail for similar reasons to Search#ModifyPaginationPortalProperties)

**Comment by john-co** [comment]
ci:test:relevant

**Comment by xbrianlee** [comment]
ci:test:search

(To catch other tests that may fail for similar reasons to Search#ModifyPaginationPortalProperties)

File: portal-web/test/functional/com/liferay/portalweb/tests/coreinfrastructure/search/searchportlet/Search.testcase [hunk 8]
```diff
@@ -364,8 +364,8 @@ definition {
 		property app.server.types = "jboss,tcserver,tomcat,weblogic,websphere,wildfly";
 		property custom.properties = "search.container.page.default.delta=2";
 		property database.types = "db2,hypersonic,mariadb,mysql,oracle,postgresql,sqlserver,sybase";
-		property environment.acceptance = "true";
-		property portal.acceptance = "true";
+		property environment.acceptance = "quarantine";
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/coreinfrastructure/search/searchportlet/Search.testcase:367]: For some of our FI owned modules, our ci relevant query also run tests that have `property environment.acceptance = "true" as well. So I set this to quarantine. Hopefully that's all that's needed to properly quarantine. Rerunning relevant tests. @zhangyan0701 can you follow up with the test results if they're not finished before my end of day and `ci:forward` if looks good?

**Comment by john-co** [comment]
ci:test:search

**Comment by john-co** [comment]
ci:forward

**Comment by suilin** [comment]
ci:test:bundle

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/run_scripts_in_element.es.js [hunk 9]
```diff
@@ -40,8 +40,8 @@ function runJSFromFile(src, next, appendFn) {
 		next();
 	};
 
-	scriptElement.addEventListener('load', callback);
-	scriptElement.addEventListener('error', callback);
+	scriptElement.addEventListener('load', callback, {once: true});
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/run_scripts_in_element.es.js:43]: I had no idea about this being a config option, thats great!

**Pull request closed without merging**